src/lib/pickles_types/shifted_value.ml

```diff
@@ -25,67 +56,135 @@ open Core_kernel
+    (module Field_intf with type t = 'f) -> shift:'f Shift.t -> 'f t -> 'f
 
-module type Field_intf = sig
-  type t
+  val equal : ('f -> 'f -> 'bool) -> 'f t -> 'f t -> 'bool
+end
 
-  val size_in_bits : int
+module Type1 = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type 'f t = Shifted_value of 'f
+      [@@deriving sexp, compare, equal, yojson, hash]
+    end
+  end]
 
-  val ( - ) : t -> t -> t
+  let typ f =
+    let there (Shifted_value x) = x in
+    let back x = Shifted_value x in
+    Snarky_backendless.Typ.(
+      transport_var (transport f ~there ~back) ~there ~back)
 
-  val ( + ) : t -> t -> t
+  let map (Shifted_value x) ~f = Shifted_value (f x)
 
-  val ( * ) : t -> t -> t
+  module Shift : sig
+    type 'f t = private { c : 'f; scale : 'f }
 
-  val inv : t -> t
+    val create : (module Field_intf with type t = 'f) -> 'f t
 
-  val one : t
+    val map : 'a t -> f:('a -> 'b) -> 'b t
+  end = struct
+    type 'f t = { c : 'f; scale : 'f }
 
-  val of_int : int -> t
-end
+    let map t ~f = { c = f t.c; scale = f t.scale }
 
-module Shift : sig
-  type 'f t = private { c : 'f; scale : 'f }
+    (* 2^{field size in bits} + 1 *)
+    let create (type f) (module F : Field_intf with type t = f) : f t =
+      { c = F.(two_to_the (module F) size_in_bits + one)
+      ; scale = F.(inv (of_int 2))
+      }
+  end
 
-  val create : (module Field_intf with type t = 'f) -> 'f t
+  let of_field (type f) (module F : Field_intf with type t = f)
+      ~(shift : f Shift.t) (s : f) : f t =
+    Shifted_value F.((s - shift.c) * shift.scale)
 
-  val map : 'a t -> f:('a -> 'b) -> 'b t
-end = struct
-  type 'f t = { c : 'f; scale : 'f }
-
-  let map t ~f = { c = f t.c; scale = f t.scale }
-
-  (* 2^{field size in bits} + 1 *)
-  let create (type f) (module F : Field_intf with type t = f) : f t =
-    let rec two_to_the n =
-      if n = 0 then F.one
-      else
-        let r = two_to_the (n - 1) in
-        F.(r + r)
-    in
-    { c = F.(two_to_the size_in_bits + one); scale = F.(inv (of_int 2)) }
+  let to_field (type f) (module F : Field_intf with type t = f)
+      ~(shift : f Shift.t) (Shifted_value t : f t) : f =
+    F.(t + t + shift.c)
+
+  let equal equal (Shifted_value t1) (Shifted_value t2) = equal t1 t2
 end
 
-let of_field (type f) (module F : Field_intf with type t = f)
-    ~(shift : f Shift.t) (s : f) : f t =
-  Shifted_value F.((s - shift.c) * shift.scale)
+(* When the scalar field is larger than the inner field of the circuit,
+   we need to encode a scalar [s] as a pair ((s >> 1), s & 1). In other
+   words, the high bits, and then the low bit separately.
+
+   We can then efficiently compute the function
 
-let to_field (type f) (module F : Field_intf with type t = f)
-    ~(shift : f Shift.t) (Shifted_value t : f t) : f =
-  F.(t + t + shift.c)
+   f = fun (g, s) -> (2 * (s >> 1) + (s & 1) + 2^(5 * ceil(len(s >> 1) / 5))) g
+     = fun (g, s) -> (s + 2^field_size_in_bits) g
```

Comment:
perhaps explain what this function is. It's not immediately clear, especially if you don't know that this is how vbmul is implemented

Reply:
also I'm not sure I understand why it uses the constant 5